dev/Kozani/KozaniRemoteManager/KozaniDvcServer.cpp

```diff
@@ -0,0 +1,734 @@
+            RETURN_HR_IF(HRESULT_FROM_WIN32(ERROR_INVALID_DATA), !isPartialData);
+            copyChunk = true;
+            break;
+
+        case CHANNEL_FLAG_LAST:
+            // Last chunk - needs to be copied and complete the Kozani PDU.
+            // isPartialData must be true to indicate that we saw and copied the first chunk already.
+            RETURN_HR_IF(HRESULT_FROM_WIN32(ERROR_INVALID_DATA), !isPartialData);
+            copyChunk = true;
+            isLastChunk = true;
+            isPartialData = false;
+            break;
+
+        default:
+            FAIL_FAST_MSG("Logic error - should have covered all cases for chunkFlags");
+        }
+
+        if (copyChunk)
+        {
+            auto startingSize = pduBuffer.size();
+            auto bytesToCopy = size - sizeof(CHANNEL_PDU_HEADER); // Skip the header.
+
+            // The partial PDU size should not be greater than the total PDU size recorded in the first data chunk.
+            RETURN_HR_IF(HRESULT_FROM_WIN32(ERROR_INVALID_DATA), startingSize + bytesToCopy > pduSize);
+
+            pduBuffer.resize(startingSize + bytesToCopy);
+            RtlCopyMemory(pduBuffer.data() + startingSize, chunkPayload, bytesToCopy);
+
+            if (isLastChunk)
+            {
+                // The completed PDU should have a size consistent with the total size of segmented data in the first chunk header.
+                RETURN_HR_IF(HRESULT_FROM_WIN32(ERROR_INVALID_DATA), pduBuffer.size() != pduSize);
+                *pdu = pduBuffer.data();
+            }
+        }
+
+        return S_OK;
+    }
+    CATCH_RETURN()
+
+    HRESULT KozaniDvcServer::ProcessProtocolDataUnit(_In_reads_(size) const BYTE* data, UINT32 size) noexcept try
+    {
+        Dvc::ProtocolDataUnit pdu;
+        if (!pdu.ParseFromArray(data, size))
+        {
+            RETURN_HR_MSG(E_FAIL, "Failed to parse PDU. size: %u.", size);
+        }
+
+        Dvc::ProtocolDataUnit::DataType type{ pdu.type() };
+        switch (type)
+        {
+            case Dvc::ProtocolDataUnit::ActivateAppRequest:
+                {
+                    // Process the ActivateAppRequest in a separate thread so a long activation will not block subsequent activation requests.
+                    auto activateAppThread{ std::thread(&KozaniDvcServer::ProcessActivateAppRequestThreadFunc, this, std::move(pdu)) };
+                    activateAppThread.detach();
+                }
+            
+                break;
+
+            case Dvc::ProtocolDataUnit::AppTerminationNotice:
+                RETURN_IF_FAILED(ProcessAppTerminationNotice(pdu));
+                break;
+
+            default:
+                // ToDo: Unsupported PDU type. Return GenericResult message with failure code to client side.
+                break;
+        }
+
+        return S_OK;
+    }
+    CATCH_RETURN()
+
+    void KozaniDvcServer::ProcessActivateAppRequestThreadFunc(_In_ Dvc::ProtocolDataUnit pdu) noexcept try
```

Comment:
This is intentional. Need a deep copy of pdu so it can be used in a different thread while the original one (which is stack allocate) gets released in the original thread.